fly/src/sorts.rs

```diff
@@ -331,27 +248,181 @@ impl UnifyValue for OptionSort {
+    /// and return the sort of the term.
+    pub fn sort_check_term(&self, term: &mut Term) -> Result<Sort, SortError> {
+        let mut unification_table = UnificationTable::new();
+        let mut internal_ctx = InternalContext::new(self.clone(), &mut unification_table);
+        // The sort inference algorithm proceeds in two phases.
+        //
+        // First, we walk the entire AST and do normal "type checking" checks.
+        // During this pass, if we encounter a bound variable without a sort
+        // annotation, we allocate a unification variable for it. The unification
+        // variable is recorded in the unification table (self.bound_names) and the bound
+        // variable in the AST is labeled with a string that uniquely identifies its
+        // corresponding variable. Other checks elsewhere in the AST (possibly above
+        // the node!) should resolve the unification variable to a concrete sort.
+        // This concrete sort gets stored in the unification table, but the AST
+        // still has the unification variable recorded as the sort of the variable.
+        //
+        // Second, we walk the AST again, looking for bound variables with
+        // unification variables. We assert that the variable was successfully
+        // resolved to a concrete sort (if not, report an error to the user that a
+        // type annotation is required), and then replace the unification variable
+        // with the concrete sort.
+
+        // Phase 1
+        let sort = internal_ctx.collect_sort_constraints_term(term)?;
+
+        // Next check if we have enough information to fully determine the sort of term.
+        // If not, error. If so, proceed to phase 2.
+        match internal_ctx.get_abstract_sort(&sort) {
+            Some(sort) => {
+                // At this point, bound variables without sort annotations have "var {id}"
+                // as their sort annotation. Second pass fixes this.
+
+                // Phase 2
+                internal_ctx.annotate_solved_sorts_term(term)?;
+                // Double check that we found all blank sort annotations in phase 1.
+                assert!(has_all_sort_annotations_term(term));
+                Ok(sort)
+            }
+            None => Err(SortError::UnsolvedSort(
+                "the term given to get_term_sort".to_owned(),
+            )),
+        }
+    }
+
+    /// Use sort inference to check that the term has the expected sort
+    pub fn sort_check_term_expect(
+        &self,
+        term: &mut Term,
+        expected: &Sort,
+    ) -> Result<(), SortError> {
+        let mut unification_table = UnificationTable::new();
+        let mut internal_ctx = InternalContext::new(self.clone(), &mut unification_table);
+        let sort = internal_ctx
+            .collect_sort_constraints_term(term)
+            .and_then(|var| {
+                internal_ctx.unify_var_value(expected, &var)?;
+                Ok(var)
+            })?;
```

Comment:
(Minor) Move the `?` up above the `and_then` to exit as early as possible. You'll have to move the `unify_var_value` cal to take the `collect_sort_constraints_term` call as an argument, and delete the `and_then`.